helm/revoke.go

```diff
@@ -0,0 +1,156 @@
+			return err
+		}
+		logger.Infof("Successfully revoked access from RBAC groups")
+	} else {
+		logger.Infof("No RBAC groups to revoke - skipping")
+	}
+
+	if len(rbacUsers) > 0 {
+		logger.Infof("Revoking access from RBAC users")
+		if err := revokeAccessFromRBACEntities(kubectlOptions, tillerNamespace, convertToGroupInfos(rbacUsers)); err != nil {
+			return err
+		}
+		logger.Infof("Successfully revoked access from RBAC users")
+	} else {
+		logger.Infof("No RBAC users to revoke - skipping")
+	}
+
+	if len(serviceAccounts) > 0 {
+		logger.Infof("Revoking access from RBAC Service Accounts")
+		serviceAccountInfos, err := convertToServiceAccountInfos(serviceAccounts)
+		if err != nil {
+			return err
+		}
+		if err := revokeAccessFromRBACEntities(kubectlOptions, tillerNamespace, serviceAccountInfos); err != nil {
+			logger.Errorf("Error revoking access from Service Accounts: %s", err)
+			return err
+		}
+		logger.Infof("Successfully revoked access from Service Accounts")
+	} else {
+		logger.Infof("No Service Accounts to revoke - skipping")
+	}
+
+	return nil
+}
+
+// revokeAccessFromRBACEntities will revoke access to the deployed Tiller server to the provided RBAC groups.
+// The revocation process is as follows:
+// For each RBAC entity:
+// - Discover and remove the role and rolebinding
+// - Discover and remove the signed TLS keypair
+func revokeAccessFromRBACEntities(
+	kubectlOptions *kubectl.KubectlOptions,
+	tillerNamespace string,
+	rbacEntities []RBACEntity,
+) error {
+	logger := logging.GetProjectLogger()
+
+	numEntities := len(rbacEntities)
+	for idx, rbacEntity := range rbacEntities {
+		logger.Infof("Revoking access to entity %s (%d of %d entities)", rbacEntity, idx+1, numEntities)
+
+		logger.Infof("Deleting entity role")
+		if err := deleteEntityRoleAndBinding(kubectlOptions, tillerNamespace, rbacEntity); err != nil {
+			logger.Errorf("Unable to delete role for entity %s from namespace %s", rbacEntity.Subject().Name, tillerNamespace)
+			return err
+		}
+
+		logger.Infof("Deleting entity keypair from secrets")
+		if err := deleteEntityTLS(kubectlOptions, tillerNamespace, rbacEntity); err != nil {
+			logger.Errorf("Unable to delete TLS keypair for entity %s from namespace %s", rbacEntity.Subject().Name, tillerNamespace)
+			return err
+		}
+	}
+	return nil
+}
+
+// deleteEntityRoleandBinding deletes the RBAC role and role binding associated with a provided entity
+func deleteEntityRoleAndBinding(
+	kubectlOptions *kubectl.KubectlOptions,
+	tillerNamespace string,
+	rbacEntity RBACEntity,
+) error {
+	logger := logging.GetProjectLogger()
+
+	client, err := kubectl.GetKubernetesClientFromOptions(kubectlOptions)
+	if err != nil {
+		return err
+	}
+
+	emptyOptions := &metav1.DeleteOptions{}
+	roleName := getTillerAccessRoleName(rbacEntity.EntityID(), tillerNamespace)
+	err = client.RbacV1().Roles(tillerNamespace).Delete(roleName, emptyOptions)
```

Comment:
I believe this will error if the object doesn't exist. Might be better to check if it exists, and if it doesn't, continue (logging a warning of course). This way, you can rerun the command for retryable errors (e.g rate limiting).

Reply:
Don't forget to do this for all the delete calls!

Reply:
Sounds good, I will update these to log warnings and move on. E.g.
```
err = client.RbacV1().Roles(tillerNamespace).Delete(roleName, emptyOptions)
if err != nil {
	logger.Warningf("Unable to delete RBAC role: %s", err)
} else {
	logger.Infof("Successfully deleted RBAC role %s", roleName)
}
```

Reply:
Oh can you actually specifically check if the object exists and only skip if the object check fails? Best would be to introspect the error and only skip on the error that is thrown when the object doesn't exist. I know this isn't easy to discover because `client-go` isn't particularly well documented, so you might have to resort to trial and error to figure this out.

I don't think I have done this anywhere else (I use `List` interfaces in the other places I need it, and that works because the objects have sufficient labels). Alternatively, you can update the grant function to attach unique labels so that we can use the `List` function interface for existence checks. I slightly prefer the latter, since that has other, user friendly benefits.

Reply:
Oh, I understand you now. I'll make this change.

Reply:
@yorinasub17 this change will make `revoke` incompatible with roles/bindings created with earlier versions of `kubergrunt` that don't have labels. Are you okay with that?

Reply:
Yes I think that is ok. It isn't ideal, but it is not the end of the world since we don't have as many customers using this right now. We can provide migration instructions as part of the release notes, which can use `kubectl` commands to delete the previous certs and then regrant using the new `helm grant`.